Add various fast-paths in typechecking

stedolan opened the pull request:
Here are a few trivial optimisations to the typechecker, which amount to skipping a complicated computation in the easy case. They add up to about 5% saved in time and allocations during typechecking (measured on typecore.ml)

  - ~~Skip Builtin_attributes.warning_scope if there are no attributes~~ (Actually, the Warnings.save/restore seems to be required by some tests. I've disabled this one, as it didn't have very much effect anyway)
  - Skip Typecore.type_unpacks if there are no first-class modules
  - Skip Rec_check if the recursive binding is a function
  - Skip computation of disambiguation warnings, if disabled

stedolan committed:
Add various fast-paths in typechecking.

  - Skip Typecore.type_unpacks if there are no first-class modules
  - Skip Rec_check if the recursive binding is a function
  - Skip computation of disambiguation warnings, if disabled
  - Make Includemod.pair_components tail recursive

gasche commented on typing/rec_check.ml:
```diff
@@ -1202,17 +1205,16 @@ and is_destructuring_pattern : type k . k general_pattern -> bool =
-                          but does not depend on rec-bound variables *)
-      true
+  match classify_expression expr with
+  | Static_Delay -> true
```
I'm not fond of this approach to implement the `Rec_check` fast path, can you explain why you did things that way?

In my mind there are two independent computations regarding recursive values:
- their "size"
- their "modes"

The final part of this module, that is the function here, combines the two analyses to make the final accept/reject decision. 

Your change introduces a new notion of size `Static_Delay` whose name mentions a mode and which means "fast path please", and is used for expressions of a certain shape. But here (in the "final" part) we do have the expression at hand, so why don't we check for functions here as a first check (and other fast-path conditions if any), then ask for the size in the non-function case, and then the modes?

(Remark: your change also contains a small refactoring where we check for the size first and then do the mode-checking differently, it does make the code nicer and it's a good approach to keep. But I have the impression that we could have this without the weird  `Static_Delay` size.)

stedolan replied:
The old code calls `classify_expression expr` (cheap) and `expression expr Return` (expensive), and case-splits on the results. The observation here is that if `classify_expression expr` classified into three rather than two cases, you could skip the expensive call to `expression expr Return` in the common case, because functions have both static size and guarantee to delay all accesses.

I have no particular opinions on how this should be done, I'm not at all familiar with this code. I'm happy with any approach as long as it avoids calling `expression expr Return` on functions.

gasche commented:
Out of curiosity: how do you measure the cost centers in the type-checker?

stedolan commented:
> Out of curiosity: how do you measure the cost centers in the type-checker?

I used a mixture of memtrace (https://github.com/janestreet/memtrace, https://blog.janestreet.com/finding-memory-leaks-with-memtrace/), and `perf` (to confirm that time-based sampling yielded roughly the same results as allocation-based sampling), and some as-yet-unreleased scripts to convert memtrace traces to a form viewable in Perfetto (https://perfetto.dev/)

If you'd like to play with memtrace, there's a sample trace of the compiler here: https://github.com/janestreet/memtrace/blob/master/examples/ocamlopt.ctf

stedolan commented:
I've updated the Rec_check code, and added one more small change: making Includemod.pair_components tail recursive, by using a match-with-exception instead of a try.

gasche commented:
Thanks! I'm happy with the new `Rec_check` variant.

I think that with more work we could speedup the `Rec_check` computation by making it return immediately in certain `Delay` contexts, even within subexpressions. (It is not obvious how to do this because we may still want to distinguish Delay and Ignore, but we don't necessarily want to pay for a free-variable computation traversal.) We could go in this direction if need arises later, but for now I think that your fast-path does capture most situations (maybe `Texp_lazy` could be added eventually).
